Add a raycast call to the webgl fat lines code

nathanieltagg opened the pull request:
I find myself in need of a reliable thick-line WebGL implementation, and this one seems the one that I can make work.. it only lacks a raycaster.  I've written my own here, and tested with private code.

nathanieltagg committed:
Add a working raycast call.

nathanieltagg committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

WestLangley commented on examples/js/lines/LineSegments2.js:
```diff
@@ -11,7 +11,7 @@ THREE.LineSegments2 = function ( geometry, material ) {
 	this.geometry = geometry !== undefined ? geometry : new THREE.LineSegmentsGeometry();
 	this.material = material !== undefined ? material : new THREE.LineMaterial( { color: Math.random() * 0xffffff } );
-
+  this.raycast_to_segment = true; // False if you don't care which segment is cast.
```
For consistency, let's not add this.

WestLangley commented on examples/js/lines/LineSegments2.js:
```diff
@@ -60,6 +60,76 @@ THREE.LineSegments2.prototype = Object.assign( Object.create( THREE.Mesh.prototy
-	}
+	}, 
+  
+  // NJT added:
```
Thanks for this, but such comments are not required.

WestLangley commented on examples/js/lines/LineSegments2.js:
```diff
@@ -60,6 +60,76 @@ THREE.LineSegments2.prototype = Object.assign( Object.create( THREE.Mesh.prototy
+			var vEnd = new THREE.Vector3();
+			var interSegment = new THREE.Vector3();
+			var interRay = new THREE.Vector3();
+			var step = ( this && this.isLineSegments ) ? 2 : 1;
```
`step` is not used.

WestLangley commented on examples/js/lines/LineSegments2.js:
```diff
@@ -60,6 +60,76 @@ THREE.LineSegments2.prototype = Object.assign( Object.create( THREE.Mesh.prototy
+      var starts = geometry.attributes.instanceStart;
+      var ends   = geometry.attributes.instanceEnd;
+      
+      for(var i=0;i<starts.count;i++) {
```
Please see https://zz85.github.io/mrdoobapproves/ for help correcting your code style.

WestLangley commented:
Thank your for your contribution. Please also test on really fat lines. I expect there will be problems, but this approximation may be acceptable.

Mugen87 commented:
@nathanieltagg It would be great if you can implement @WestLangley's feedback since users in the forum requested this feature multiple times:

https://discourse.threejs.org/t/raycasting-with-linesegment2-objects/4256

nathanieltagg committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

nathanieltagg committed:
Add PerspectiveLineMaterial, update example

nathanieltagg committed:
Add the PerspectiveLineMaterial

nathanieltagg commented:
Hi, I've responded to the code changes as requested.

I've decided to add one more enhancement, a new material called PerspectiveLineMaterial.  This draws the line with a thickness that varies with the distance from the camera. This is a nice way of drawing simple frame structures without needing full 3d tubes. I've been using this in my own code.

I've also updated the fatlines demo to preview the new material, and added a raycast marker to show that it works.

nathanieltagg commented:
(Sorry if I've done something wrong; this is my first pull request ever.)

nathanieltagg commented:
OK, I think I've fixed it. I've also added a new feature I think is pretty useful: a perspective line.

— Nathaniel





> On May 8, 2019, at 1:43 PM, Michael Herzog <notifications@github.com> wrote:
> 
> @nathanieltagg <https://github.com/nathanieltagg> It would be great if you can implement @WestLangley <https://github.com/WestLangley>'s feedback since users in the forum requested this feature multiple times:
> 
> https://discourse.threejs.org/t/raycasting-with-linesegment2-objects/4256 <https://discourse.threejs.org/t/raycasting-with-linesegment2-objects/4256>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/mrdoob/three.js/pull/16240#issuecomment-490581869>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAQJNBPY3E77FDOISQPOXELPUMGNHANCNFSM4HFZDRDQ>.
>
